src/index.js

```diff
@@ -2,39 +2,44 @@ import ReactDOM from 'react-dom'
 import {Simulate} from 'react-dom/test-utils'
 import {getQueriesForElement, prettyDOM} from 'dom-testing-library'
 
-function render(
-  ui,
-  {container = document.createElement('div'), baseElement = container} = {},
-) {
+const mountedContainers = new Set()
+
+function render(ui, {container, baseElement = container} = {}) {
+  if (!container) {
+    baseElement = document.documentElement
+    container = document.body.appendChild(document.createElement('div'))
+  }
+
+  // we'll add it to the mounted containers regardless of whether it's actually
+  // added to document.body so the cleanup method works regardless of whether
+  // they're passing us a custom container or not.
+  mountedContainers.add(container)
+
   ReactDOM.render(ui, container)
   return {
     container,
     // eslint-disable-next-line no-console
-    debug: () => console.log(prettyDOM(baseElement)),
+    debug: (el = baseElement) => console.log(prettyDOM(el)),
     unmount: () => ReactDOM.unmountComponentAtNode(container),
     rerender: rerenderUi => {
-      render(rerenderUi, {container})
+      render(rerenderUi, {container, baseElement})
       // Intentionally do not return anything to avoid unnecessarily complicating the API.
       // folks can use all the same utilities we return in the first place that are bound to the container
     },
     ...getQueriesForElement(baseElement),
   }
 }
 
-const mountedContainers = new Set()
-
-function renderIntoDocument(ui) {
-  const container = document.body.appendChild(document.createElement('div'))
-  mountedContainers.add(container)
-  return render(ui, {container, baseElement: document.documentElement})
+function cleanup() {
+  mountedContainers.forEach(cleanupAtContainer)
 }
 
-function cleanup() {
-  mountedContainers.forEach(container => {
-    document.body.removeChild(container)
-    ReactDOM.unmountComponentAtNode(container)
-    mountedContainers.delete(container)
-  })
+// maybe one day we'll expose this (perhaps even as a utility returned by render).
+// but let's wait until someone asks for it.
+function cleanupAtContainer(container) {
+  document.body.removeChild(container)
```

Comment:
If `container` is custom passed by the caller, thus not appended to the `document.body` by `render`, the caller may not expect that it will be removed by `cleanup` which should be a mirror to `render`.

Reply:
Hmmm... So what are you suggesting?

Reply:
Oh, I see what you mean. I'm not sure how to solve that issue reliably. If they don't like the behavior then they can file an issue and we can add an option to prevent this. Let's wait to add complexity until someone has an issue.

Reply:
Usually if a piece of code owns the lifecycle of an object (the caller in the case of custom container; the module with the render function in the case of default container), that piece is responsible for both creation+setup and teardown+destruction.

The suggestion is to make sure not to remove the container from the DOM if it hasn't been added to the DOM by the render module, only remove the map entry (lifecycle of which is owned by the render module) and `ReactDOM.unmountComponentAtNode` (which is a mirror to `ReactDOM.render`, mounted component's lifecycle is also owned by the render module).

I understand that the object/memory ownership and the allocate/deallocate pattern is often missed/violated in the JavaScript community due to various reasons, so it may look odd to require following it from this library users.

Reply:
That makes sense. Perhaps we should not add items to the set if we don't append it to the body. Would you like to open a PR for that?

Reply:
We still need to keep a reference to unmount it, we could use a `Map<HTMLElement,boolean>` instead of `Set<HTMLElement>` to store whether the container was created (owned) or borrowed.

Reply:
Hmmmm..... I'm still not certain this is the behavior people would expect our want 🤔

Reply:
What if instead we did:

```javascript
container.parentElement && container.parentElement.removeChild(container)
```

I think that would make sense to happen in the cleanup regardless of whether it's in the body or not.

Reply:
That would make sense to shorten the caller code to let it manage only `container` creation, and, my point still holds, the ownership of the `container` object lifecycle is moved from the caller into the `render` function.

Reply:
I'm a little confused by what you're suggesting. But why don't you make a PR and maybe that'll clear things up for me.